### ISSUE TITLE
1sec pause on disconnect 

### ISSUE THREAD

**Issue by fpagliughi** [eclipse-paho/paho.mqtt.c#889]
I'm working on a demo system that uses a sat phone modem to make quick connections in order to get the data home:

- Turn on the modem
- Connect
- Publish one or two messages (QoS=1)
- Wait for completion (CONNACK)
- Disconnect
- Turn off the modem

For reasons of billing and electrical power usage (this is battery powered), this needs to happen as fast as possible.

The problem is that there is usually (but not always) a 1sec pause in the disconnection. I'm assuming it must be hitting on a wait for something to complete, but the wait seems pretty long.

**Comment by icraggs** [comment]
Is it waiting for the pubacks?  A trace might be the easiest way to answer that and other questions.

**Comment by fpagliughi** [comment]
No, I don't believe so. I'm calling disconnect _after_ publish call(s) complete.

And, just to be clear, it does this on a test setup over localhost.  I'm just looking to address the issue before deploying to the sat phone. I'll post a trace.

**Comment by fpagliughi** [comment]
Interesting... it doesn't seem to do it when I turn tracing on. I'll sniff the packets to make sure it's not something in my system. But the fairly precise 1-sec timing of the delay had me convinced that it was a sleep() call somewhere in the disconnect process.

**Comment by landersson** [comment]
I remember also noticing a similar delay when disconnecting, while trying to speed up integration tests. Did any of you find the cause?

**Comment by icraggs** [comment]
I haven't found anything.  @fpagliughi ?

**Comment by fpagliughi** [comment]
Oh, I had forgotten about this. Give me a few days to look into it again and I will report back.

**Comment by fpagliughi** [comment]
This is something like what I was talking about. Here it seems like a substantial pause, not in the disconnect, but in `MQTTAsync_destroy()`. This sounds familiar. Did we discuss it in a different issue recently?

This log is from a C++ sample app that connects using TCP on localhost, publishes a few messages, then disconnects. 
https://github.com/eclipse/paho.mqtt.cpp/blob/master/src/samples/async_publish.cpp

It's using C v1.3.6 on x86_64 Linux (Mint 19).

```
...
Sending final message...
20201023 231240.583 3 async_publish -> PUBLISH qos: 0 retained: 0 rc: 0 payload len(28): Someone is always li
20201023 231240.583 Calling publish success for client async_publish
	Listener success for token: 0
OK

Disconnecting...
20201023 231240.583 3 async_publish -> DISCONNECT (0)
20201023 231240.583 Removed socket 3
20201023 231240.583 Reset max fdp1 to 1
20201023 231240.583 0 inflight messages deleted for client async_publish
20201023 231240.583 0 responses removed for client async_publish
20201023 231240.583 Calling disconnect complete for client async_publish
  ...OK
20201023 231240.583 0 inflight messages deleted for client async_publish
20201023 231240.583 0 responses removed for client async_publish
20201023 231240.583 0 responses removed for client async_publish
20201023 231240.583 0 commands removed for client async_publish
20201023 231240.583 Removed client async_publish from bstate->clients, socket 0
20201023 231240.583 Conn_count is 0
20201023 231240.583 sleeping
20201023 231240.583 sleeping
20201023 231240.583 sleeping
20201023 231240.583 sleeping
20201023 231240.583 sleeping
20201023 231240.583 sleeping    <-|
20201023 231241.285 sleeping    <-|   ~700ms
20201023 231241.285 sleeping
20201023 231241.285 sleeping
20201023 231241.285 Return code 0 from read select
20201023 231241.285 Return code 0 from write select
```

**Comment by fpagliughi** [comment]
Oh, yeah, is that it... in `MQTTAsync_stop()` ?

```
while ((sendThread_state != STOPPED || receiveThread_state != STOPPED) && ++count < 100)
{
	MQTTAsync_unlock_mutex(mqttasync_mutex);
	Log(TRACE_MIN, -1, "sleeping");
	MQTTAsync_sleep(100L);
	MQTTAsync_lock_mutex(mqttasync_mutex);
}
```

That's probably what I was seeing.
Now... what's keeping the threads from stopping?

**Comment by icraggs** [comment]
There's a PR which might be relevant: #983

**Comment by fpagliughi** [comment]
I don't think it's the same thing. In the C++ publish sample, it sends a few QoS=1 messages and waits for them to all be acknowledged before disconnecting. For some odd reason in the sample it even does a double check to make sure there are no messages still in flight:
```
// Double check that there are no pending tokens

auto toks = client.get_pending_delivery_tokens();
if (!toks.empty())
    cout << "Error: There are pending delivery tokens!" << endl;

// Disconnect
conntok = client.disconnect();
conntok->wait();
```
So before the disconnect and stop, all activity has completed. That's why I was assuming the shutdown would happen very quickly.

**Comment by shashankanayak** [comment]
Hi @icraggs @fpagliughi 
I see the same issue MQTTAsync_destroy is taking 1 second due  MQTTAsync_receiveThread which has  select timeout of 1000L.(1sec)
Can we fix this issue, in our embedded device this is considerable delay,which is bringing latency between our microservices.
micoservice-a---->microservice b
we are using request reply pattern of paho c++ library
where  we are getting a roundtrip delay of 1sec.[[Delay is from the c library as pointed above not in the transport]](since the communication is on same system we expect the delay to be less than 100ms )[[usbus and msgq gives o.1millisecond latency]]

**Comment by icraggs** [comment]
@shashankanayak if you can reproduce this frequently (which it sounds like) you could take a client library trace to show what happens.   On destroy might be a different problem however.

**Comment by shashankanayak** [comment]
@icraggs 
attaching logs
[log.txt](https://github.com/eclipse/paho.mqtt.c/files/6547750/log.txt)

**Comment by icraggs** [comment]
You'll need a trace without PAHO_HIGH_PERFORMANCE, because that removes most of the trace entries.  And just the normal level, not maximum.

**Comment by shashankanayak** [comment]
Hi @icraggs ,
As suggested please find new set of logs attached.with log level with performance flag disabled.
export MQTT_C_CLIENT_TRACE_LEVEL=PROTOCOL,MAXIMUM,MEDIUM attached.
Issue is reproducible all the time.
[log_maximum.txt](https://github.com/eclipse/paho.mqtt.c/files/6560149/log_maximum.txt)
[log_protcol.txt](https://github.com/eclipse/paho.mqtt.c/files/6560150/log_protcol.txt)
[log_medium.txt](https://github.com/eclipse/paho.mqtt.c/files/6560151/log_medium.txt)

**Comment by shashankanayak** [comment]
hi @icraggs ,
did you have chance to look into the logs?

**Comment by icraggs** [comment]
In changing the select call to poll, which can be interrupted by a socket being closed locally, I think this issue is fixed, in the develop branch.

**Comment by fpagliughi** [comment]
After all the nice performance changes to v1.3.16, I'm seeing this issue again. I can't verify if this was truly fixed years ago, or if I just wasn't noticing it recently. 

Again, like with connect, I'm wondering if this is due to the 1sec sleep in one of the threads, preventing it from checking for disconnect?

**Comment by fpagliughi** [comment]
I ran a simple publisher example app in a loop.  Each run of the application connects to Mosquitto on localhos, publishes a single small messages with QoS=1, and then disconnects (with a zero timeout). My guess is that it should take < 10ms total.

Running it over and over takes the following times to complete. This was measured as the "real" time duration from the Linux `time` utility. It seems like there may be both a 1000ms and a 100ms sleep that are blocking the disconnect.

```
1.109s
0.112s
1.011s
0.109s
1.008s
1.113s
1.008s
1.015s
0.112s
0.111s
0.110s
1.013s
...
```

**Comment by fpagliughi** [comment]
Actually, it seems (with the current issue at least) that the problem is not with the disconnect, but with the stop/destroy. It waits for the send and receive threads to complete, and they are each potentially blocked in a 1sec sleep.

Signaling both of these seems to fix the issue. PR coming shortly.

**Comment by icraggs** [comment]
That's a nice way to fix this issue.
